add  kwarg to handle invalid files in open_mfdataset

pratiman-91 opened the pull request:
- [x] Closes #6736 
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

Added new argument in `open_mfdataset` to better handle the invalid files.
```
errors : {'ignore', 'raise', 'warn'}, default 'raise'
        - If 'raise', then invalid dataset will raise an exception.
        - If 'ignore', then invalid dataset will be ignored.
        - If 'warn', then a warning will be issued for each invalid dataset.
```

max-sixty commented:
I'm not the expert, but this looks reasonable! Any other thoughts?

Assuming no one thinks it's a bad idea, we would need tests.

headtr1ck commented on xarray/backends/api.py:
```diff
@@ -1393,6 +1394,7 @@ def open_mfdataset(
     join: JoinOptions = "outer",
     attrs_file: str | os.PathLike | None = None,
     combine_attrs: CombineAttrsOptions = "override",
+    errors: str = "raise",
```
```suggestion
    errors: ErrorOptionsWithWarn = "raise",
```
From `xarray.core.types`

pratiman-91 replied:
Thanks @headtr1ck I did not know about `xarray.core.type`.

headtr1ck commented on xarray/backends/api.py:
```diff
@@ -1611,7 +1617,36 @@ def open_mfdataset(
+                    )
+                    continue
+        else:
+            raise ValueError(
```
This is never being raised because it is catched by the outer try-except.

Also, to keep the complexity lower, check this before entering the if else block: `if errors not in ("raise", "warn", "ignore"): raise ValueError(...)`

pratiman-91 replied:
Thanks @headtr1ck for pointing this out. I have made changes to the logic based on your suggestions.

headtr1ck reviewed (changes requested):
I think it is a good idea.

But the way it is implemented here seems overly complicated and repetitive.
I would suggest to revert the logic: first build up the list wrapped in a single `try` and then handle the three cases in the `except` block.

headtr1ck commented on xarray/backends/api.py:
```diff
@@ -1611,7 +1617,27 @@ def open_mfdataset(
         getattr_ = getattr
 
-    datasets = [open_(p, **open_kwargs) for p in paths1d]
+    if errors in ("raise", "warn", "ignore"):
```
```suggestion
    if errors not in ("raise", "warn", "ignore"):
        raise ValueError(f"'errors' must be 'raise', 'warn' or 'ignore', got '{errors}'")
```
And then de-dent everything below.

headtr1ck commented on xarray/backends/api.py:
```diff
@@ -1611,7 +1617,27 @@ def open_mfdataset(
+            except Exception:
+                if errors == "raise":
+                    raise
+                elif errors == "ignore":
```
```suggestion
                elif errors == "warn":
```

headtr1ck reviewed (commented):
Almost there.

Also, we should add tests for this.

pratiman-91 commented:
@headtr1ck Thanks for the suggestions. I have added two tests (`ignore` and `warn`). Also, while testing, I found that a new argument broke `combine="nested"` due to invalid `ids`. I have now modified it to reflect the correct `ids`, and it is passing the tests. Please review the tests and the latest version.

headtr1ck commented on xarray/backends/api.py:
```diff
@@ -1626,6 +1654,7 @@ def open_mfdataset(
         if combine == "nested":
             # Combined nested list by successive concat and merge operations
             # along each dimension, using structure given by "ids"
+            ids = [id_ for i, id_ in enumerate(ids) if i not in invalid_ids]
```
Does this work with nested datasets? E.g. a 2x2 set where one file is missing

pratiman-91 replied:
Thanks for pointing this out. I did not know about 2x2 functionality. Now, I have modified the logic to remove the invalid ids and paths and added tests for the same.

pratiman-91 commented:
Hi @headtr1ck, I have been thinking about the handling of `ids`. Current version looks like a patch work (I am not happy with it.). I think we can create ids after removing all the invalid datasets from `path1d` within the `combine==nested` block. Please let me know what do you think. 
Thanks!

pratiman-91 commented:
@max-sixty Can you please go through the PR. Thanks!

max-sixty commented on xarray/backends/api.py:
```diff
@@ -1368,6 +1369,38 @@ def open_groups(
     return groups
 
 
+def remove_path(paths, path_to_remove):
```
maybe make this private? and add return type to emphasize that we're just returning a new list rather than removing paths from disk...

pratiman-91 replied:
done!

pratiman-91 commented:
It seems like one test failed `test_sparse_dask_dataset_repr (xarray.tests.test_sparse.TestSparseDataArrayAndDataset) `. It is not related to this PR.

pratiman-91 commented:
@dcherian @max-sixty Can you please help me merge this request? I am not sure what more needs to be done.

Also, every day I am getting this message. Not sure how to resolve this. 
![image](https://github.com/user-attachments/assets/0cc82719-1649-440f-9590-5177aeb8fa65)

max-sixty commented:
can we fix the errors in the docs?

pratiman-91 commented:
I just rebased the branch. Let us see if this resolves the docs, as I have not made any changes in the docs.

pratiman-91 commented:
> can we fix the errors in the docs?

I just rebased the branch. Let us see if this resolves the docs, as I have not made any changes in the docs.

pratiman-91 commented:
@max-sixty all tests passed.

kmuehlbauer commented on xarray/backends/api.py:
```diff
@@ -1540,6 +1560,12 @@ def open_mfdataset(
         If a callable, it must expect a sequence of ``attrs`` dicts and a context object
         as its only parameters.
+    errors : {"raise", "warn", "ignore"}, default: "raise"
+        String indicating how to handle errors in opeining dataset.
```
```suggestion
        String indicating how to handle errors in opening dataset.
```

kmuehlbauer commented on xarray/backends/api.py:
```diff
@@ -1540,6 +1560,12 @@ def open_mfdataset(
+
+        - "raise": invalid dataset will raise an exception.
+        - "warn": a warning will be issued for each invalid dataset.
+        - "ignore": then invalid dataset will be ignored.
```
```suggestion
        - "ignore": invalid dataset will be ignored.
```

kmuehlbauer commented on xarray/backends/api.py:
```diff
@@ -1632,7 +1658,32 @@ def open_mfdataset(
+                warnings.warn(
+                    f"Could not open {p}. Ignoring.", UserWarning, stacklevel=2
+                )
+                continue
+            else:
+                continue
```
I'm not sure, but the complete else-branch is redundant and probably the first continue, too?

pratiman-91 replied:
I have simplified it to `if`.

kmuehlbauer commented on xarray/backends/api.py:
```diff
@@ -1632,7 +1658,32 @@ def open_mfdataset(
+        try:
+            ds = open_(p, **open_kwargs)
+            datasets.append(ds)
+        except Exception:
```
Can we catch the exception and give the user a more descriptive warning in case of "warn" why the open failed?

pratiman-91 replied:
that is a good idea. thanks for the suggestion.

kmuehlbauer reviewed (changes requested):
@pratiman-91 Thanks for sticking with this. This seams all reasonable to me. I've added a couple comments and suggestions. 

For the tests, I think it would be great to also test for "raise".

pratiman-91 commented:
@kmuehlbauer Thank you for the suggestions and typos. I have made changes based on your suggestions. Please check. 

For the tests, I did not write a test for "raise" since it is the default and already being tested by various other tests. If you think it would be beneficial, I can write another test for "raise".

kmuehlbauer commented on xarray/backends/api.py:
```diff
@@ -1632,7 +1658,32 @@ def open_mfdataset(
+        try:
+            ds = open_(p, **open_kwargs)
+            datasets.append(ds)
+        except Exception as e:
+            # remove invalid paths
+            if combine == "nested":
+                paths = _remove_path(paths, p)
+                remove_paths = True
+            if errors == "raise":
+                raise
+            if errors == "warn":
+                warnings.warn(
+                    f"Could not open {p} due to {e}. Ignoring.",
+                    UserWarning,
+                    stacklevel=2,
+                )
+            continue
```
I think this would be a bit less involved. We can use `emit_user_level_warning` here, too.


```suggestion
        except Exception as e:
            if errors == "raise":
                raise
            elif errors == "warn":
                emit_user_level_warnings(f"Could not open {p} due to {e}. Ignoring.")
            # remove invalid paths
            if combine == "nested":
                paths = _remove_path(paths, p)
                remove_paths = True
```

kmuehlbauer commented:
@pratiman-91 I've thought a bit about the removing of the problematic files for the nested case.

To my understanding we have two corresponding 1D lists (`ids` and `paths1d`).

Wouldn't this work to just remember the indices fro the failing files and remove those from the existing lists afterwards? Along the lines:

```python
remove = []
for i, paths in enumerate(paths1d):
   try:
       open(...)
   except:
      if combine="nested":
          remove.append(i)
          
# later in combine nested

if remove_paths:
    for i in sorted(remove, reverse=True):
        del ids[i]
        del paths1d[i]
    
```

pratiman-91 commented:
I have thought about this approach before. However, it creates a problem with a 2x2 nested case. Therefore, I have settled for recreating the IDs and paths.

kmuehlbauer commented:
Thanks for the pointer @pratiman-91. Yes, this has a twist. 

I see this is a special case for the nested combine. All our good faith doesn't help, for these cases. If the user provides a 2x2, or NxM nested list or whatever and one or more files are missing in one or more of the sublists the whole thing explodes. So I'd rather go for having "raise" in those cases as having the user see `ValueError: The supplied objects do not form a hypercube because sub-lists do not have consistent lengths along dimension X`.

Maybe others have some opinion here?

pratiman-91 commented on xarray/tests/test_backends.py:
```diff
@@ -5014,6 +5014,68 @@ def test_open_mfdataset_2(self) -> None:
+            original.isel(x=slice(5), y=slice(4, 8)).to_netcdf(tmp3)
+            original.isel(x=slice(5, 10), y=slice(4, 8)).to_netcdf(tmp4)
+            with open_mfdataset(
+                [[tmp1, tmp2], ["non-existent-file.nc", tmp3, tmp4]],
```
You are correct @kmuehlbauer, removing `tmp3` will cause the test to fail. The current implementation expects the input to align with the intended concat/merge structure. However, if a file is missing or invalid and its removal still results in a valid nested list structure, then the process should still work as expected. Otherwise,  `raise` an error to indicate an inconsistency in the input and concat/merge not possible.

pratiman-91 commented:
@kmuehlbauer @max-sixty Any update on this PR? 
Thanks!

-Pratiman

kmuehlbauer reviewed (changes requested):
@pratiman-91 This is almost there. The whats-new entry is missing. Probably got lost at some point in time.

kmuehlbauer commented:
@headtr1ck Would you mind having one last look here? I'm not able to merge this without your interaction wrt to a requested change. Thanks!

kmuehlbauer commented:
Thanks @pratiman-91 for sticking with us and congrats to your first contribution!
